File: android/src/main/java/com/retenosdk/RetenoEcomEvent.java
package com.retenosdk;

import static com.reteno.core.domain.model.ecom.EcomEvent.CartUpdated;
import static com.reteno.core.domain.model.ecom.EcomEvent.OrderCancelled;
import static com.reteno.core.domain.model.ecom.EcomEvent.OrderCreated;
import static com.reteno.core.domain.model.ecom.EcomEvent.OrderDelivered;
import static com.reteno.core.domain.model.ecom.EcomEvent.OrderUpdated;
import static com.reteno.core.domain.model.ecom.EcomEvent.ProductAddedToWishlist;
import static com.reteno.core.domain.model.ecom.EcomEvent.ProductCategoryViewed;
import static com.reteno.core.domain.model.ecom.EcomEvent.ProductViewed;
import static com.reteno.core.domain.model.ecom.EcomEvent.SearchRequest;

import android.util.Pair;

import com.facebook.react.bridge.ReadableArray;
import com.facebook.react.bridge.ReadableMap;
import com.reteno.core.domain.model.ecom.Attributes;
import com.facebook.react.bridge.ReadableType;
import com.reteno.core.domain.model.ecom.Order;
import com.reteno.core.domain.model.ecom.OrderItem;
import com.reteno.core.domain.model.ecom.OrderStatus;
import com.reteno.core.domain.model.ecom.ProductCategoryView;
import com.reteno.core.domain.model.ecom.ProductInCart;
import com.reteno.core.domain.model.ecom.ProductView;

import java.time.ZonedDateTime;
import java.util.ArrayList;

import javax.annotation.Nullable;

public class RetenoEcomEvent {

  @Nullable
  static Attributes buildAttributesFromPayload(ReadableMap payload) {
    if (payload == null) return null;
    String name = RetenoUtil.getStringOrNull(payload.getString("name"));
    ReadableArray value = payload.getArray("value");
    ArrayList<String> values = new ArrayList<>();
    if (value != null) {
      for (int i = 0; i < value.size(); i++) {
        values.add(value.getString(i));
      }
    }
    if (name == null) return null;

    return new Attributes(
      name,
      values
    );
  }

  @Nullable
  static Pair<String, String> buildSimpleAttributesFromPayload(ReadableMap payload) {
    if (payload == null) return null;
    String name = RetenoUtil.getStringOrNull(payload.getString("name"));
    String value = RetenoUtil.getStringOrNull(payload.getString("value"));
    if (name == null || value == null) return null;

    return new Pair<>(
      name,
      value
    );
  }

  @Nullable
  static ProductView buildProductFromPayload(ReadableMap payload) {

    if (payload == null) return null;
    String productId = payload.getString("productId");
    if (productId == null) return null;
    double price = payload.getDouble("price");
    boolean isInStock = payload.getBoolean("isInStock");
    ReadableArray attributes = payload.getArray("attributes");
    ArrayList<Attributes> attrList = new ArrayList<>();
    Attributes attr;

    if (attributes != null) {
      for (int i = 0; i < attributes.size(); i++) {
        attr = buildAttributesFromPayload(attributes.getMap(i));
        if (attr != null) {
          attrList.add(attr);
        }
      }
    }

    return new ProductView(
      productId,
      price,
      isInStock,
      attrList
    );
  }

  @Nullable
  static ProductViewed buildProductViewedFromPayload(ReadableMap payload) {
    if (payload == null) return null;
    ProductView productView = buildProductFromPayload(payload.getMap("product"));
    if (productView == null) return null;
    String currencyCode = RetenoUtil.getStringOrNull(payload.getString("currencyCode"));

    return new ProductViewed(
      productView,
      currencyCode
    );
  }

  @Nullable
  public static ProductCategoryViewed buildProductCategoryViewedFromPayload(ReadableMap payload) {
    if (payload == null) return null;
    ProductCategoryView categoryView = buildProductCategoryFromPayload(payload.getMap("category"));
    if (categoryView == null) return null;

    return new ProductCategoryViewed(categoryView);
  }

  @Nullable
  private static ProductCategoryView buildProductCategoryFromPayload(ReadableMap payload) {
    if (payload == null) return null;
    String productCategoryId = payload.getString("productCategoryId");
    if (productCategoryId == null) return null;
    ReadableArray attributesArray = payload.getArray("attributes");
    ArrayList<Attributes> attributesList = null;
    Attributes attr;

    if (attributesArray != null) {
      attributesList = new ArrayList<>();
      for (int i = 0; i < attributesArray.size(); i++) {
        attr = buildAttributesFromPayload(attributesArray.getMap(i));
        if (attr != null) {
          attributesList.add(attr);
        }
      }
    }

    return new ProductCategoryView(
      productCategoryId,
      attributesList
    );
  }

  @Nullable
  public static ProductAddedToWishlist buildProductAddedToWishlistFromPayload(ReadableMap payload) {
    if (payload == null) return null;
    ProductView productView = buildProductFromPayload(payload.getMap("product"));
    if (productView == null) return null;
    String currencyCode = payload.getString("currencyCode");

    return new ProductAddedToWishlist(
      productView,
      currencyCode
    );
  }

  @Nullable
  public static CartUpdated buildCartUpdatedFromPayload(ReadableMap payload) {
    if (payload == null) return null;
    String cartId = payload.getString("cartId");
    if (cartId == null) return null;
    ReadableArray products = payload.getArray("cartItems");
    if (products == null) return null;
    ArrayList<ProductInCart> productList = new ArrayList<>();

    for (int i = 0; i < products.size(); i++) {
      ProductInCart item = buildCartItemFromPayload(products.getMap(i));
      if (item != null) {
        productList.add(item);
      }
    }

    String currencyCode = payload.getString("currencyCode");

    return new CartUpdated(
      cartId,
      productList,
      currencyCode
    );
  }

  @Nullable
  static ProductInCart buildCartItemFromPayload(ReadableMap payload) {
    if (payload == null) return null;
    String productId = payload.getString("productId");
    if (productId == null) return null;
    int quantity = payload.getInt("quantity");
    double price = payload.getDouble("price");
    Double discount = payload.getDouble("discount");
    String name = payload.getString("name");
    String category = payload.getString("category");

    ReadableArray attributesArray = payload.getArray("attributes");
    ArrayList<Attributes> attributesList = null;
    Attributes attr;

    if (attributesArray != null) {
      attributesList = new ArrayList<>();
      for (int i = 0; i < attributesArray.size(); i++) {
        attr = buildAttributesFromPayload(attributesArray.getMap(i));
        if (attr != null) {
          attributesList.add(attr);
        }
      }
    }

    return new ProductInCart(
      productId,
      quantity,
      price,
      discount,
      name,
      category,
      attributesList
    );
  }

  @Nullable
  public static OrderCreated buildOrderCreatedFromPayload(ReadableMap payload) {
    if (payload == null) return null;
    Order order = buildOrderFromPayload(payload.getMap("order"));
    if (order == null) return null;
    String currencyCode = payload.getString("currencyCode");

    return new OrderCreated(
      order,
      currencyCode
    );
  }

  private static Double getSafeDouble(ReadableMap payload, String key) {
    if (payload.getType(key) == ReadableType.Number) {
    
    return payload.getDouble(key);
  } else {
    return null;
  }
}

@Nullable
  private static Order buildOrderFromPayload(ReadableMap payload) {
    if (payload == null) return null;
    String externalOrderId = payload.getString("externalOrderId");
    if (externalOrderId == null) return null;
    String externalCustomerId = payload.getString("externalCustomerId");
    double totalCost = payload.getDouble("totalCost");
    OrderStatus status = OrderStatus.values()[payload.getInt("status")];
    String cartId = payload.getString("cartId");
    String email = payload.getString("email");
    String phone = payload.getString("phone");
    String name = payload.getString("firstName");
    String lastName = payload.getString("lastName");
    Double shipping = getSafeDouble(payload, "shipping");
    Double discount = getSafeDouble(payload, "discount");
    Double taxes = getSafeDouble(payload, "taxes");
    String restoreId = payload.getString("restoreId");
    String statusDescription = payload.getString("statusDescription");
    String storeId = payload.getString("storeId");
    String source = payload.getString("source");
    String deliveryMethod = payload.getString("deliveryMethod");
    String paymentMethod = payload.getString("paymentMethod");
    String deliveryAddress = payload.getString("deliveryAddress");
    ReadableArray orderItems = payload.getArray("orderItems");
    ArrayList<OrderItem> orderItemList = null;
    if (orderItems != null) {
      orderItemList = new ArrayList<>();
      for (int i = 0; i < orderItems.size(); i++) {
        OrderItem item = buildOrderItemFromPayload(orderItems.getMap(i));
        if (item != null) {
          orderItemList.add(item);
        }
      }
    }


    ReadableArray attributes = payload.getArray("attributes");
    ArrayList<Pair<String, String>> attrPairs = null;

    if (attributes != null) {
      attrPairs = new ArrayList<>();
      for (int i = 0; i < attributes.size(); i++) {
        Pair<String, String> pair = buildSimpleAttributesFromPayload(attributes.getMap(i));
        if (pair != null) {
          attrPairs.add(pair);
        }
      }
    }


    return new Order(
      externalOrderId,
      externalCustomerId,
      totalCost,
      status,
      ZonedDateTime.now(),
      cartId,
      email,
      phone,
      name,
      lastName,
      shipping,
      discount,
      taxes,
      restoreId,
      statusDescription,
      storeId,
      source,
      deliveryMethod,
      paymentMethod,
      deliveryAddress,
      orderItemList,
      attrPairs
    );
  }

  @Nullable
  static OrderItem buildOrderItemFromPayload(ReadableMap payload) {
    if (payload == null) return null;
    String externalItemId = payload.getString("externalItemId");
    if (externalItemId == null) return null;
    String name = payload.getString("name");
    if (name == null) return null;
    String category = payload.getString("category");
    if (category == null) return null;
    int quantity = payload.getInt("quantity");
    double price = payload.getDouble("price");
    String url = payload.getString("url");
    if (url == null) return null;
    String imageUrl = payload.getString("imageUrl");
    String description = payload.getString("description");


    return new OrderItem(
      externalItemId,
      name,
      category,
      quantity,
      price,
      url,
      imageUrl,
      description
    );
  }

  @Nullable
  public static OrderUpdated buildOrderUpdatedFromPayload(ReadableMap payload) {
    if (payload == null) return null;
    Order order = buildOrderFromPayload(payload.getMap("order"));
    if (order == null) return null;
    String currencyCode = payload.getString("currencyCode");

    return new OrderUpdated(
      order,
      currencyCode
    );
  }

  @Nullable
  public static OrderDelivered buildOrderDeliveredFromPayload(ReadableMap payload) {
    if (payload == null) return null;
    String externalOrderId = payload.getString("externalOrderId");
    if (externalOrderId == null) return null;

    return new OrderDelivered(
      externalOrderId
    );
  }

  @Nullable
  public static OrderCancelled buildOrderCancelledFromPayload(ReadableMap payload) {
    if (payload == null) return null;
    String externalOrderId = payload.getString("externalOrderId");
    if (externalOrderId == null) return null;

    return new OrderCancelled(
      externalOrderId
    );
  }

  @Nullable
  public static SearchRequest buildSearchRequestFromPayload(ReadableMap payload) {
    if (payload == null) return null;
    String search = payload.getString("searchQuery");
    if (search == null) return null;
    boolean isFound = payload.getBoolean("isFound");


    return new SearchRequest(
      search,
      isFound
    );
  }
}

File: ios/RetenoEcomEvent.swift
import Foundation
import Reteno

@objc(RetenoEcomEvent)
class RetenoEcomEvent: NSObject {
    
    static func buildProductDataFromPayload(_ payload: [String: Any]?) -> (product: Ecommerce.Product, currencyCode: String?)? {
        guard let payload = payload else { return nil }
        guard let productDict = payload["product"] as? [String: Any],
              let product = buildProductFromPayload(productDict) else { return nil }
        let currencyCode = payload["currencyCode"] as? String
        
        return (product, currencyCode)
    }
    
    static func buildProductCategoryDataFromPayload(_ payload: [String: Any]?) -> Ecommerce.ProductCategory? {
        guard let payload else { return nil }
        guard let categoryDict = payload["category"] as? [String: Any],
              let category = buildProductCategoryFromPayload(categoryDict) else { return nil }
        
        return category
    }
    
    static func buildCartUpdatedDataFromPayload(_ payload: [String: Any]?) -> (cartId: String, products: [Ecommerce.ProductInCart], currencyCode: String?)? {
        guard let payload = payload else { return nil }
        guard let cartId = payload["cartId"] as? String else { return nil }
        guard let productsArray = payload["cartItems"] as? [[String: Any]] else { return nil }
        
        var productList: [Ecommerce.ProductInCart] = []
        for productDict in productsArray {
            if let item = buildCartItemFromPayload(productDict) {
                productList.append(item)
            }
        }
        
        let currencyCode = payload["currencyCode"] as? String
        
        return (
            cartId: cartId,
            products: productList,
            currencyCode: currencyCode
        )
    }
    
    static func buildOrderDataFromPayload(_ payload: [String: Any]?) -> (order: Ecommerce.Order, currencyCode: String?)? {
        guard let payload = payload else { return nil }
        guard let orderDict = payload["order"] as? [String: Any],
              let order = buildOrderFromPayload(orderDict) else { return nil }        

        let currencyCode = payload["currencyCode"] as? String
        
        return (
            order: order,
            currencyCode: currencyCode
        )
    }
    
    static func buildOrderExternalIdFromPayload(_ payload: [String: Any]?) -> String? {
        guard let payload = payload else { return nil }
        guard let externalOrderId = payload["externalOrderId"] as? String else { return nil }
        
        return externalOrderId
    }
    
    static func buildSearchRequestDataFromPayload(_ payload: [String: Any]?) -> (searchQuery: String, isFound: Bool)? {
        guard let payload = payload else { return nil }
        guard let searchQuery = payload["searchQuery"] as? String else { return nil }
        let isFound = payload["isFound"] as? Bool ?? false
        
        return (
            searchQuery: searchQuery,
            isFound: isFound
        )
    }

    private static func buildAttributesFromPayload(_ payload: [String: Any]?) -> [String: [String]]? {
        guard let payload = payload else { return nil }
        guard let name = payload["name"] as? String else { return nil }
        guard let valueArray = payload["value"] as? [String] else { return nil }
        return [name: valueArray]
    }
    
    private static func buildProductFromPayload(_ payload: [String: Any]?) -> Ecommerce.Product? {
        guard let payload = payload else { return nil }
        guard let productId = payload["productId"] as? String else { return nil }
        guard let price = payload["price"] as? Float else { return nil }
        let isInStock = payload["isInStock"] as? Bool ?? false
        
        var attributes: [String: [String]] = [:]
        if let attributesArray = payload["attributes"] as? [[String: Any]] {
            for attr in attributesArray {
                if let attribute = buildAttributesFromPayload(attr) {
                    attributes.merge(attribute) { (_, new) in new }
                }
            }
        }
        
        return Ecommerce.Product(
            productId: productId,
            price: price,
            isInStock: isInStock,
            attributes: attributes
        )
    }
    
    private static func buildProductCategoryFromPayload(_ payload: [String: Any]?) -> Ecommerce.ProductCategory? {
        guard let payload = payload else { return nil }
        guard let productCategoryId = payload["productCategoryId"] as? String else { return nil }
        
        var attributesList: [String: [String]] = [:]
        if let attributesArray = payload["attributes"] as? [[String: Any]] {
            for attr in attributesArray {
                if let attribute = buildAttributesFromPayload(attr) {
                    attributesList.merge(attribute) { (_, new) in new }
                }
            }
        }
        
        return Ecommerce.ProductCategory(
            productCategoryId: productCategoryId,
            attributes: attributesList
        )
    }
    
    private static func buildCartItemFromPayload(_ payload: [String: Any]?) -> Ecommerce.ProductInCart? {
        guard let payload = payload else { return nil }
        guard let productId = payload["productId"] as? String else { return nil }
        guard let quantity = payload["quantity"] as? Int else { return nil }
        guard let price = payload["price"] as? Float else { return nil }
        let discount = payload["discount"] as? Float
        let name = payload["name"] as? String
        let category = payload["category"] as? String
        
        var attributesList: [String: [String]] = [:]
        if let attributesArray = payload["attributes"] as? [[String: Any]] {
            for attr in attributesArray {
                if let attribute = buildAttributesFromPayload(attr) {
                    attributesList.merge(attribute) { (_, new) in new }
                }
            }
        }
        
        return Ecommerce.ProductInCart(
            productId: productId,
            price: price,
            quantity: quantity,
            discount: discount,
            name: name,
            category: category,
            attributes: attributesList
        )
    }
    
    private static func buildOrderItemFromPayload(_ payload: [String: Any]?) -> Ecommerce.Order.Item? {
        guard let payload = payload else { return nil }
        guard let externalItemId = payload["externalItemId"] as? String else { return nil }
        guard let name = payload["name"] as? String else { return nil }
        guard let category = payload["category"] as? String else { return nil }
        guard let quantity = payload["quantity"] as? Double else { return nil }
        guard let price = payload["price"] as? Float else { return nil }
        guard let url = payload["url"] as? String else { return nil }
        
        let imageUrl = payload["imageUrl"] as? String
        let description = payload["description"] as? String
        
        return Ecommerce.Order.Item(
            externalItemId: externalItemId,
            name: name,
            category: category,
            quantity: quantity,
            cost: price,
            url: url,
            imageUrl: imageUrl,
            description: description
        )
    }


    
    private static func buildOrderFromPayload(_ payload: [String: Any]?) -> Ecommerce.Order? {

              enum LocalStatus: Int {
                case initialized = 1
                case inprogress = 2
                case delivered = 3
                case cancelled = 4
                
                var toOrderStatus: Ecommerce.Order.Status {
                    switch self {
                    case .initialized:
                        return Ecommerce.Order.Status(rawValue: "INITIALIZED")!
                    case .inprogress:
                        return Ecommerce.Order.Status(rawValue: "INPROGRESS")!
                    case .delivered:
                        return Ecommerce.Order.Status(rawValue: "DELIVERED")!
                    case .cancelled:
                        return Ecommerce.Order.Status(rawValue: "CANCELLED")!
                    }
                }
            }


        guard let payload = payload else { return nil }
        guard let externalOrderId = payload["externalOrderId"] as? String else { return nil }
        
        let externalCustomerId = payload["externalCustomerId"] as? String
        guard let totalCost = payload["totalCost"] as? Float else { return nil }
        
        // guard let statusRaw = payload["status"] as? String,
        //       let status = StatusStatus(rawValue: statusRaw) else { return nil }
           guard let statusRaw = payload["status"] as? Int,
                 let localStatus = LocalStatus(rawValue: statusRaw) else { return nil }
    
                 let orderStatus = localStatus.toOrderStatus

       
        let cartId = payload["cartId"] as? String
        let email = payload["email"] as? String
        let phone = payload["phone"] as? String
        let firstName = payload["firstName"] as? String
        let lastName = payload["lastName"] as? String
        let shipping = payload["shipping"] as? Float ?? 0.0
        let discount = payload["discount"] as? Float ?? 0.0
        let taxes = payload["taxes"] as? Float ?? 0.0
        let restoreId = payload["restoreId"] as? String
        let statusDescription = payload["statusDescription"] as? String
        let storeId = payload["storeId"] as? String
        let source = payload["source"] as? String
        let deliveryMethod = payload["deliveryMethod"] as? String
        let paymentMethod = payload["paymentMethod"] as? String
        let deliveryAddress = payload["deliveryAddress"] as? String
        
        var orderItemList: [Ecommerce.Order.Item]?
        if let orderItems = payload["orderItems"] as? [[String: Any]] {
            orderItemList = []
            for itemDict in orderItems {
                if let item = buildOrderItemFromPayload(itemDict) {
                    orderItemList?.append(item)
                }
            }
        }
        
        let attributePairs = payload["attributes"] as? [String: [String: Any]]


        return Ecommerce.Order(
            externalOrderId: externalOrderId,
            totalCost: totalCost,
            status: orderStatus,
            date: Date(),
            cartId: cartId,
            email: email,
            phone: phone,
            firstName: firstName,
            lastName: lastName,
            shipping: shipping,
            discount: discount,
            taxes: taxes,
            restoreUrl: restoreId,
            statusDescription: statusDescription,
            storeId: storeId,
            source: source,
            deliveryMethod: deliveryMethod,
            paymentMethod: paymentMethod,
            deliveryAddress: deliveryAddress,
            items: orderItemList,
            attributes: attributePairs
        )
    }
}
